language-plutus-core/src/Language/PlutusCore/Error.hs

```diff
@@ -122,7 +134,7 @@ instance AsRenameError (Error a) a where
 instance AsTypeError (Error a) a where
     _TypeError = _TypeErrorE
 
-instance AsNormalizationError (Error a) TyName Name a where
+instance (tyname ~ TyName, name ~ Name) => AsNormalizationError (Error a) tyname name a where
```

Comment:
Why make this change, OOI?

Reply:
In `Renamer.hs` we have

```haskell
instance (Coercible unique1 Unique, unique1 ~ unique2) =>
        HasUniquesRenaming (UniquesRenaming unique1) unique2 where
```

which we could make

```haskell
instance (Coercible unique Unique) =>
        HasUniquesRenaming (UniquesRenaming unique) unique where
```

but then our code stops to type check.

it can be the case that somewhere we have code that requires type annotations while it shouldn't. So, no particular reason, just an easy change to help type inference.

Reply:
Cool, I would have assumed that they were the same. Are equality constraints generally better for type inference than concrete types in the head?

Reply:
Yes, but that doesn't come free, of course. The new instance

```haskell
instance (tyname ~ TyName, name ~ Name) => AsNormalizationError (Error a) tyname name a where
```

says that a `NormalizationError tyname name a` can be treated as an `Error a` as long as `tyname` is `TyName` and `name` is `Name`, which is exactly what we want in this case, but often it's just too restrictive and we want to allow other instantiations of `tyname` and `name` to be possible.

[Related blog post](https://chrisdone.com/posts/haskell-constraint-trick)

Reply:
I'm actually probably wrong and in this particular case equality constraints do not affect anything. The generated class is

```haskell
class AsNormalizationError r
                           (tyname :: * -> *)
                           (name :: * -> *)
                           a | r -> tyname name a where
```

I.e. we already can't define instances with different `tyname`s and `name`s, because the functional dependency says that `tyname` and `name` are determined by `r`.

But I'm not really sure if the equality constraints affect anything or not.

Reply:
Maybe it matters if you're trying to do something with overlapping instances, then I think the one with the explicit types in the head will be more specific. Anyway, seems fine.